From: Eric Dumazet <edumazet@google.com>
Date: Sun, 29 Sep 2013 01:21:32 -0700
Subject: [PATCH 146/242] net: skb_is_gso_v6() requires skb_is_gso()
Patch-mainline: v3.13-rc1

Git-commit: 36a8f39e05ccc308a5619a7edb5ad6e15ee82ff6 (partial)
References: bsc#908684 FATE#317539


bnx2x makes a dangerous use of skb_is_gso_v6().

It should first make sure skb is a gso packet

Signed-off-by: Eric Dumazet <edumazet@google.com>
Cc: Eilon Greenstein <eilong@broadcom.com>
Acked-by: Dmitry Kravkov <dmitry@broadcom.com>
Signed-off-by: David S. Miller <davem@davemloft.net>
Acked-by: Cho, Yu-Chen <acho@suse.com>
---
 drivers/net/ethernet/broadcom/bnx2x/bnx2x_cmn.c |   10 ++++++----
 1 file changed, 6 insertions(+), 4 deletions(-)

--- a/drivers/net/ethernet/broadcom/bnx2x/bnx2x_cmn.c
+++ b/drivers/net/ethernet/broadcom/bnx2x/bnx2x_cmn.c
@@ -3169,10 +3169,12 @@ static inline u32 bnx2x_xmit_type(struct
 		}
 	}
 
-	if (skb_is_gso_v6(skb))
-		rc |= XMIT_GSO_V6 | XMIT_CSUM_TCP | XMIT_CSUM_V6;
-	else if (skb_is_gso(skb))
-		rc |= XMIT_GSO_V4 | XMIT_CSUM_V4 | XMIT_CSUM_TCP;
+	if (skb_is_gso(skb)) {
+		if (skb_is_gso_v6(skb))
+			rc |= XMIT_GSO_V6 | XMIT_CSUM_TCP | XMIT_CSUM_V6;
+		else 
+			rc |= XMIT_GSO_V4 | XMIT_CSUM_V4 | XMIT_CSUM_TCP;
+	}
 
 	return rc;
 }
